File: qr_reader.go
// qr_reader.go - Ридер QR-кодов на Go (CLI)
// Установка: go get -u github.com/makiuchi-d/gozxing
package main

import (
	"encoding/csv"
	"flag"
	"fmt"
	"image"
	_ "image/gif"
	_ "image/jpeg"
	_ "image/png"
	"os"
	"path/filepath"
	"strings"

	"github.com/makiuchi-d/gozxing"
	"github.com/makiuchi-d/gozxing/qrcode"
)

func main() {
	var imagePath, batchPath, outputPath string
	flag.StringVar(&imagePath, "image", "", "Путь к изображению")
	flag.StringVar(&batchPath, "batch", "", "Папка для пакетного сканирования")
	flag.StringVar(&outputPath, "output", "", "Файл для сохранения результатов (CSV)")
	flag.Parse()

	if batchPath != "" {
		batchScan(batchPath, outputPath)
	} else if imagePath != "" {
		decodeImage(imagePath)
	} else {
		fmt.Println("Укажите --image или --batch")
	}
}

func decodeImage(path string) {
	file, err := os.Open(path)
	if err != nil {
		fmt.Printf("Ошибка открытия %s: %v\n", path, err)
		return
	}
	defer file.Close()

	img, _, err := image.Decode(file)
	if err != nil {
		fmt.Printf("Ошибка декодирования %s: %v\n", path, err)
		return
	}

	bmp, err := gozxing.NewBinaryBitmapFromImage(img)
	if err != nil {
		fmt.Printf("Ошибка создания битмапа: %v\n", err)
		return
	}

	qrReader := qrcode.NewQRCodeReader()
	result, err := qrReader.Decode(bmp, nil)
	if err != nil {
		fmt.Println("QR-код не найден")
		return
	}

	fmt.Printf("📷 QR-код распознан!\n")
	fmt.Printf("Данные: %s\n", result.GetText())
	fmt.Printf("Формат: %s\n", result.GetBarcodeFormat())
}

func batchScan(folder, output string) {
	entries, err := os.ReadDir(folder)
	if err != nil {
		fmt.Printf("Ошибка чтения папки: %v\n", err)
		return
	}

	var results [][]string
	extensions := map[string]bool{".png": true, ".jpg": true, ".jpeg": true, ".bmp": true, ".gif": true}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		ext := strings.ToLower(filepath.Ext(entry.Name()))
		if !extensions[ext] {
			continue
		}
		fullPath := filepath.Join(folder, entry.Name())
		file, err := os.Open(fullPath)
		if err != nil {
			continue
		}
		img, _, err := image.Decode(file)
		file.Close()
		if err != nil {
			continue
		}
		bmp, err := gozxing.NewBinaryBitmapFromImage(img)
		if err != nil {
			continue
		}
		qrReader := qrcode.NewQRCodeReader()
		result, err := qrReader.Decode(bmp, nil)
		if err == nil {
			results = append(results, []string{entry.Name(), result.GetText()})
			fmt.Printf("%s: %s\n", entry.Name(), result.GetText())
		}
	}

	if output != "" && len(results) > 0 {
		f, err := os.Create(output)
		if err != nil {
			fmt.Printf("Ошибка создания файла: %v\n", err)
			return
		}
		defer f.Close()
		writer := csv.NewWriter(f)
		writer.Write([]string{"file", "data"})
		writer.WriteAll(results)
		writer.Flush()
		fmt.Printf("Результаты сохранены в %s\n", output)
	}
}

File: qr_reader.rs
// qr_reader.rs - Ридер QR-кодов на Rust (CLI)
// Зависимости: rqrr, image, clap
use clap::{Arg, App};
use image::{GenericImageView, ImageFormat};
use rqrr::PreparedImage;
use std::fs;
use std::path::Path;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let matches = App::new("QR Reader")
        .arg(Arg::with_name("image").short("i").long("image").takes_value(true).help("Путь к изображению"))
        .arg(Arg::with_name("batch").short("b").long("batch").takes_value(true).help("Папка для пакетного сканирования"))
        .arg(Arg::with_name("output").short("o").long("output").takes_value(true).help("Файл для сохранения результатов"))
        .get_matches();

    if let Some(batch) = matches.value_of("batch") {
        batch_scan(batch, matches.value_of("output"))?;
    } else if let Some(image) = matches.value_of("image") {
        decode_image(image)?;
    } else {
        println!("Укажите --image или --batch");
    }
    Ok(())
}

fn decode_image(path: &str) -> Result<(), Box<dyn std::error::Error>> {
    let img = image::open(path)?;
    let (width, height) = img.dimensions();
    let gray = img.to_luma8();

    let prepared = PreparedImage::prepare(gray);
    let results = prepared.detect_codes()?;

    if results.is_empty() {
        println!("QR-код не найден");
        return Ok(());
    }

    for result in results {
        println!("📷 QR-код распознан!");
        println!("Данные: {}", result.data());
    }
    Ok(())
}

fn batch_scan(folder: &str, output: Option<&str>) -> Result<(), Box<dyn std::error::Error>> {
    let extensions = [".png", ".jpg", ".jpeg", ".bmp", ".gif"];
    let mut results = Vec::new();

    for entry in fs::read_dir(folder)? {
        let entry = entry?;
        let path = entry.path();
        if !path.is_file() {
            continue;
        }
        let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("");
        let ext = format!(".{}", ext.to_lowercase());
        if !extensions.contains(&ext.as_str()) {
            continue;
        }

        if let Ok(img) = image::open(&path) {
            let gray = img.to_luma8();
            let prepared = PreparedImage::prepare(gray);
            if let Ok(codes) = prepared.detect_codes() {
                for code in codes {
                    let name = path.file_name().unwrap().to_string_lossy();
                    println!("{}: {}", name, code.data());
                    results.push((name.to_string(), code.data().to_string()));
                }
            }
        }
    }

    if let Some(out) = output {
        use std::io::Write;
        let mut file = fs::File::create(out)?;
        writeln!(file, "file,data")?;
        for (name, data) in results {
            writeln!(file, "{},\"{}\"", name, data)?;
        }
        println!("Результаты сохранены в {}", out);
    }
    Ok(())
}
